### DISCUSSION TITLE
Relationships to agents / BMR's made-up rule

### DISCUSSION THREAD

**Discussion by briesenberg07** [uwlib-cams/uwlswd#49] [modeling]
# Modelling question

BMR stated that they thought the following rule should apply to UWLSWD. 

***IF*** we wish to create a relationship between a resource and *one* creator (or contributor, or publisher, etc.)--say for example the University of Washington Libraries--***THEN*** we should allow ✔️  data which looks like this...
```
<> dc:creator "University of Washington Libraries"@en .
```
...or this ✔️ ...
```
<> dct:creator <http://viaf.org/viaf/143991097> .
```
...but disallow ❌ data that looks like this:
```
<> dc:creator "University of Washington Libraries"@en ;
   dct:creator <http://viaf.org/viaf/143991097> .
```
This is because, according to BMR, the last example (❌) indicates relationships between two (2) *distinct* creators. 

In contrast, TG asserts that there are no RDF 'rules' which indicate anything like this. That is, that while the triples...
```
<> dc:creator "University of Washington Libraries"@en ;
   dct:creator <http://viaf.org/viaf/143991097> .
```
...indicate that resource `<>` has `dct:creator` http://viaf.org/viaf/143991097 and `dc:creator` "University of Washington Libraries"@en, they say *nothing whatsoever* about whether http://viaf.org/viaf/143991097 and "University of Washington Libraries"@en identify two different things or the same thing.

# Effects of modelling
Compare the difficulty of outputting desired schema.org and DataCite XML in three possible scenarios:

## 1. Most permissive modelling: Allow for providing *both* an IRI and a literal value for each related agent in RDF, or *only* an IRI *or* a literal:
```
<> dc:contributor "University of Washington Libraries"@en ; # UWL
   dct:contributor <http://viaf.org/viaf/143991097> ; # UWL
   dct:contributor <http://viaf.org/viaf/137156173> . # BNF
```
### Requirements for data conversion:
a. RDF-to-DataCite:
   - If multiple values given for the same relationship (property) determine which (if any) are the same agent (very difficult)
   - If IRI value acquire label for DataCite XML

b. RDF-to-schema.org
   - If multiple agents given for the same relationship (property) determine which (if any) are the same agent
   - If literal value acquire IRI for schema.org

## 2. Less permissive modelling: Require that *only one* triple creating a relationship to an agent with a given role be created in RDF, but allow either IRI or literal values:
```
<> dc:contributor "University of Washington Libraries"@en ; # UWL
   dct:contributor <http://viaf.org/viaf/137156173> . #BNF 
```
### Requirements for data conversion:
a. RDF-to-DataCite
   - If IRI value acquire label for DataCite XML

b. RDF-to-schema.org
   - If literal value acquire IRI for schema.org
## 3. Least permissive modelling: Require that  *only one* triple creating a relationship to an agent with a given role be created in RDF, require that this be an IRI:
```
<> dct:contributor <http://viaf.org/viaf/143991097> ; # UWL
   dct:contributor <http://viaf.org/viaf/137156173> . # BNF
```
### Requirements for data conversion:
a. RDF-to-DataCite
   - Acquire a label for each given IRI for DataCite XML

b. RDF-to-schema.org
   - Output IRI as value for schema:contributor (or other relationship)

**Comment by briesenberg07** [comment]
I suppose this could really apply to relationships to things besides agents as well...

**Comment by gerontakos** [comment]
Other options:
A bnode.
Similar: RDF bag, seq, alt, or collection.

A pitfall: the actual most permissive: 
<> dc:contributor "University of Washington Libraries"@en ;
dc:contributor <http://viaf.org/viaf/143991097> .
Nothing wrong with that, we would just want to prohibit it locally.

**Comment by gerontakos** [comment]
If we require IRIs in the RDF/XML, we'll have to make a rule that we create an agent/resource in our uwlswd data when we only have a literal. In our agents data, right?

However, it's not reasonable to think we can easily retrieve labels from external data. Every dataset has a different way of doing things; the code cannot always predict the data model used by an external dataset. 

If we want to support DataCite's requirement for a literal value, I think our best option is option 1 with required literal values, IRIs optional but required if known.

We could also require a local IRI for every agent in our local data. Then we could always use IRIs in the RDF (best practice) and reliably retrieve labels when we need them (provided we have an application profile for describing agents). That is not the easiest solution, but it is the "correct" solution, right?

However, bnodes are a valid RDF instrument. We can use a bnode in the RDF/XML and all the derived serializations. That would allow us to satisfy Benjamin's rule ("creator" triples, quantity=n, means there are n creators).

**Reply by briesenberg07** [reply]
Thank you @gerontakos ! What do you and @cspayne think of the proposal 🎏 below?

**COMMENTS**

> However, it's not reasonable to think we can easily retrieve labels from external data. Every dataset has a different way of doing things; the code cannot always predict the data model used by an external dataset.

*I agree, I wouldn't want to attempt to build this into the workflow.*

> If we want to support DataCite's requirement for a literal value, I think our best option is option 1 with required literal values, IRIs optional but required if known.

*I want to support DataCite's requirement for literal values for Publisher, Creator, Contributor **and** schema.org's requirement for IRI values.*

> However, bnodes are a valid RDF instrument. We can use a bnode in the RDF/XML and all the derived serializations. That would allow us to satisfy Benjamin's rule ("creator" triples, quantity=n, means there are n creators).

*I guess I've become anti-bnode, because I find I don't like this option... Although I also considered it...*

🎏 **PROPOSAL** - Following on TG's comments:

> If we require IRIs in the RDF/XML, we'll have to make a rule that we create an agent/resource in our uwlswd data when we only have a literal. In our agents data, right?
> We could also require a local IRI for every agent in our local data. Then we could always use IRIs in the RDF (best practice) and reliably retrieve labels when we need them (provided we have an application profile for describing agents). That is not the easiest solution, but it is the "correct" solution, right?

1. To do now
   - Get list of creators/contributors/publishers for UWLSWD top-level resources (things that are identified by UWLSWD IRIs with no hash fragment identifier) 
   - Implement a label/IRI 'mapping' in XML/XSLT/Python for use in the workflow (sinopia_maps workflows do something somewhat akin to this with various references from code to [get_prop_sets.xml](https://github.com/uwlib-cams/map_storage/blob/main/xml/get_prop_sets.xml))
   - This can allow output of desired DataCite and schema.org whether RDF/XML includes literal value, IRI value, or both values, for a given contributor role - (*note that it would still require constraining labels and IRIs to those in the 'mapping'*)
2. To do later 
   - Add these creators/contributors/publishers for UWLSWD top-level resources to the [Instances of the Agent Class](https://doi.org/10.6069/uwlib.55.A.3.6) dataset
   - Update workflow to replace the step 1 'mapping' with retrieval of either a label or an IRI from our data as needed via HTTP

**Comment by gerontakos** [comment]
That's a direction I favor. However, I would rather see the list compiled (just somehow query all existing datasets to retrieve the triples we think are pertinent -- we'll probably miss a few) then, instead of a 'mapping' in XML, get the values into our instances of agents. Alternatively we can get those agents into wikidata, which would make more sense, no? and might be easier to do even. What I'm advocating for: let's keep it all in RDF environments as much as possible. Of course we would have a app profile for wikidata and could safely predict the location of labels.

**Reply by briesenberg07** [reply]
Much of this is what I'm proposing in 2 above... but I don't think an update to the agents dataset would be something that would happen quickly.

Thus step 1 first, which I'm imagining would be quite trivial for Cypress (I could probably even do it!), and would allow us to output rdf/xml-to-datacite-to-schema in the meantime...

**Reply by cspayne** [reply]
Yes, I don't think that would take too long to do.

**Reply by gerontakos** [reply]
But why not produce a list, David can produce some wikidata items, and the code can include a permanent solution. There should be very few creators/contributors/publishers, and they're probably already in wikidata already. That seems like a quick fix to me. Editing UW agents is more complicated. We could have David do it this week. As long as we can extract from wikidata.

**Reply by briesenberg07** [reply]
A Wikidata solution! I have no objections if it works for @cspayne !

**Reply by cspayne** [reply]
I'm not familiar with what the process would be to retrieve that data from Wikidata, but as long as it's possible that would be a fine solution!

**Reply by briesenberg07** [reply]
Right. I'm not either. 
I realized after posting my last comment that I don't understand exactly how this would work. That is, retrieving from Wikidata either a label-from-URI (for RDF/XML-to-DataCite XML) or URI-from-label (for RDF/XML-to-schema.org). I'm also remembering the fact that Wikidata labels can and do change at any time.
So, this is all bringing me back to [my proposal above](https://github.com/uwlib-cams/uwlswd/discussions/49#discussioncomment-7062494). But maybe @gerontakos has something in mind that I just don't understand yet?

**Comment by cspayne** [comment]
> To do now
> 
>Get list of creators/contributors/publishers for UWLSWD top-level resources (things that are identified by UWLSWD IRIs with no hash fragment identifier)
>Implement a label/IRI 'mapping' in XML/XSLT/Python for use in the workflow (sinopia_maps workflows do something somewhat akin to this with various references from code to [get_prop_sets.xml](https://github.com/uwlib-cams/map_storage/blob/main/xml/get_prop_sets.xml))
>This can allow output of desired DataCite and schema.org whether RDF/XML includes literal value, IRI value, or both values, for a given contributor role - (note that it would still require constraining labels and IRIs to those in the 'mapping')

Question: Are we still requiring that _every_ creator (publisher, contributor, etc) has a dc:creator literal value? If so, the mapping can rely on this. Otherwise, I believe we are matching both literals and IRIs - is there a neat way in xsl to avoid producing doubles if there is both a literal and an IRI of the same creator?

**Comment by gerontakos** [comment]
I don't think we decided. The idea of every-agent-as-a-literal was, if I remember correctly, to enable us to ignore IRIs in both datacite and schema.org metadata (where only literals would be used) and have the IRIs (but also the literals) in the RDF/XML and derivative serializations. If we just input literals, could we then perform a lookup, insert only IRIs in the RDF family (or literals, if an IRI is not available), literals in the datacite, and IRIs (if available) in the schema.org? In that case, no mistakes permitted in the RDF/XML literals!

**Reply by briesenberg07** [reply]
> Are we still requiring that every creator (publisher, contributor, etc) has a dc:creator literal value? If so, the mapping can rely on this. Otherwise, I believe we are matching both literals and IRIs - is there a neat way in xsl to avoid producing doubles if there is both a literal and an IRI of the same creator?

👀  **Here is what I was thinking. I'm just now realizing that this does *not* match what's in [the spreadsheet](https://docs.google.com/spreadsheets/d/1jKYPFzhE_iNsqrekX696_wFehq9NTM4v212J41jGv0s/edit?usp=sharing), so I may be mixed up. Is the below correct, or wrong?**

**RDF/XML**
- Use dct:creator / dct:contributor / dct:publisher with IRI value
- In the event that no IRI available for one of these agents, use dc:creator / dc:contributor / dc:publisher with literal value
- Observe BMR's made-up rule - N triples with dc/dct:creator = N creator agents, N triples with dc/dct:contributor = N contributor agents, and so on

**DataCite**
- Use temporary 'mapping' to match IRI-to-literal for output to DataCite

**schema.org**
- Use temporary 'mapping' to output IRIs and literals to schema.org
- In the event that agents are identified with literal, not IRI in RDF/XML, ~use temporary 'mapping' to match literal-to-IRI for output to schema~ output literal to schema.org
   - This ^^^ didn't make sense--if an IRI wasn't available at time of creating RDF/XML, it won't be available for creating schema.org

**Reply by briesenberg07** [reply]
**Addendum: Proposed future workflow**


**RDF/XML**
- Use dct:creator / dct:contributor / dct:publisher with IRI value taken from Instances of the Agent Class
- Observe BMR's made-up rule - N triples with dc/dct:creator = N creator agents, N triples with dc/dct:contributor = N contributor agents, and so on

**DataCite**
- Retrieve literal-from-IRI from Instances of the Agent class (we can do this because we can confidently anticipate the structure of our data!)

**schema.org**
- Output IRIs and literals to schema.org

**Reply by cspayne** [reply]
This makes sense to me!

**Reply by briesenberg07** [reply]
Thanks @cspayne .  
I should note that this is based on my understanding that DataCite XML wants strings and schema.org wants IRIs... understanding which is coming from my (untrustworthy) memory and not from any actual recent review of documentation... Am I right about this?

**Reply by cspayne** [reply]
It is correct that DataCite wants strings. Schema.org is up to us! But if we have both strings and IRIs available, both can be included.

**Reply by briesenberg07** [reply]
I'll amend the proposed future workflow as follows:

> ~Output IRIs to schema.org~ Output IRIs *and literals* to schema.org

**Reply by briesenberg07** [reply]
Please see updates to
- [proposed current workflow](https://github.com/uwlib-cams/uwlswd/discussions/49#discussioncomment-7115359)
- [proposed future workflow](https://github.com/uwlib-cams/uwlswd/discussions/49#discussioncomment-7115388)

**Reply by cspayne** [reply]
For our current (temporary) solution, I think the best option would be an xml file that stores the agent label and URI. 
I put together a draft that looks something like this: 
```xml
<agents
    xmlns:schema="https://schema.org/"
    xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <agent>
        <schema:name>University of Washington Libraries</schema:name>
        <schema:sameAs rdf:resource="http://viaf.org/viaf/139541794"/>
    </agent>
</agents>
```
I based this off of our sinopia_maps map_storage xml files. 
Is it okay that the root does not have a namespace, or is there a better option?

**Reply by briesenberg07** [reply]
XML instance with the needed data makes sense to me. I like it because it would be *one place* for aligning all agent IRIs/labels as needed (creator, contributor, publisher, etc.).  
(I believe that) stylesheets and/or other code would reference this to retrieve needed info during rdfxml-to-DataCite-to-schema. In future we would replace these references with references/retrieval from Instances of the Agent Class.
I don't see any need for a namespace document as this is not intended to last very long.

**Comment by briesenberg07** [comment]
@briesenberg07 look at this again - does it need to be closed?
